crates/ruff_python_formatter/src/module/mod.rs

```diff
@@ -1,38 +1,36 @@
+use crate::context::PyFormatContext;
+use crate::{AsFormat, IntoFormat, PyFormatter};
+use ruff_formatter::Format;
+use ruff_formatter::{FormatOwnedWithRule, FormatRefWithRule, FormatResult, FormatRule};
+use rustpython_parser::ast::Mod;
+
 pub(crate) mod mod_expression;
 pub(crate) mod mod_function_type;
 pub(crate) mod mod_interactive;
 pub(crate) mod mod_module;
 
-use crate::context::PyFormatContext;
-use ruff_formatter::format_element::tag::VerbatimKind;
-use ruff_formatter::prelude::*;
-use ruff_formatter::write;
-use rustpython_parser::ast::{Mod, Ranged};
+#[derive(Default)]
+pub struct FormatMod;
 
-pub(crate) struct FormatModule<'a> {
-    module: &'a Mod,
+impl FormatRule<Mod, PyFormatContext<'_>> for FormatMod {
+    fn fmt(&self, item: &Mod, f: &mut PyFormatter) -> FormatResult<()> {
+        match item {
+            Mod::Module(x) => x.format().fmt(f),
+            Mod::Interactive(x) => x.format().fmt(f),
+            Mod::Expression(x) => x.format().fmt(f),
+            Mod::FunctionType(x) => x.format().fmt(f),
+        }
+    }
 }
-
-impl<'a> FormatModule<'a> {
-    pub(crate) fn new(module: &'a Mod) -> Self {
-        Self { module }
+impl<'ast> AsFormat<PyFormatContext<'ast>> for Mod {
+    type Format<'a> = FormatRefWithRule<'a, Mod, FormatMod, PyFormatContext<'ast>>;
+    fn format(&self) -> Self::Format<'_> {
+        FormatRefWithRule::new(self, FormatMod::default())
     }
 }
-
-impl Format<PyFormatContext<'_>> for FormatModule<'_> {
-    fn fmt(&self, f: &mut Formatter<PyFormatContext<'_>>) -> FormatResult<()> {
-        let range = self.module.range();
-
-        write!(f, [source_position(range.start())])?;
-
-        f.write_element(FormatElement::Tag(Tag::StartVerbatim(
-            VerbatimKind::Verbatim {
-                length: range.len(),
-            },
-        )))?;
-        write!(f, [source_text_slice(range, ContainsNewlines::Detect)])?;
-        f.write_element(FormatElement::Tag(Tag::EndVerbatim))?;
-
-        write!(f, [source_position(range.end())])
+impl<'ast> IntoFormat<PyFormatContext<'ast>> for Mod {
+    type Format = FormatOwnedWithRule<Mod, FormatMod, PyFormatContext<'ast>>;
+    fn into_format(self) -> Self::Format {
+        FormatOwnedWithRule::new(self, FormatMod::default())
```

Comment:
Nit: Should we move this into the `generated.rs` or what's the reasoning that this is inside the module?

Reply:
This impl is a bit special since we can't use `FormatNodeRule` (not a Node), so `FormatRule` is different. I think it makes more sense to keep it here than to special case in the generation